public/scripts/macros.js

```diff
@@ -201,56 +201,42 @@ function diceRollReplace(input, invalidRollPlaceholder = '') {
 /**
  * Substitutes {{macro}} parameters in a string.
  * @param {string} content - The string to substitute parameters in.
- * @param {*} _name1 - The name of the user.
- * @param {*} _name2 - The name of the character.
- * @param {*} _original - The original message for {{original}} substitution.
- * @param {*} _group - The group members list for {{group}} substitution.
- * @param {boolean} _replaceCharacterCard - Whether to replace character card macros.
+ * @param {Object<string, *>} env - Map of macro names to the values they'll be substituted with. If the param
+ * values are functions, those functions will be called and their return values are used.
  * @returns {string} The string with substituted parameters.
  */
-export function evaluateMacros(content, _name1, _name2, _original, _group, _replaceCharacterCard = true) {
+export function evaluateMacros(content, env) {
     if (!content) {
         return '';
     }
 
-    // Replace {{original}} with the original message
-    // Note: only replace the first instance of {{original}}
-    // This will hopefully prevent the abuse
-    if (typeof _original === 'string') {
-        content = content.replace(/{{original}}/i, _original);
-    }
     content = diceRollReplace(content);
     content = replaceInstructMacros(content);
     content = replaceVariableMacros(content);
     content = content.replace(/{{newline}}/gi, '\n');
     content = content.replace(/{{input}}/gi, String($('#send_textarea').val()));
 
-    if (_replaceCharacterCard) {
-        const fields = getCharacterCardFields();
-        content = content.replace(/{{charPrompt}}/gi, fields.system || '');
-        content = content.replace(/{{charJailbreak}}/gi, fields.jailbreak || '');
-        content = content.replace(/{{description}}/gi, fields.description || '');
-        content = content.replace(/{{personality}}/gi, fields.personality || '');
-        content = content.replace(/{{scenario}}/gi, fields.scenario || '');
-        content = content.replace(/{{persona}}/gi, fields.persona || '');
-        content = content.replace(/{{mesExamples}}/gi, fields.mesExamples || '');
+    // Substitute passed-in variables
+    for (const varName in env) {
+        if (!Object.hasOwn(env, varName)) continue;
```

Comment:
`substituteParams` sets `user` and `char` first, so they'll always be substituted first since object keys are [iterated over in insertion order unless they're numbers](https://stackoverflow.com/a/5525820).

I can add a comment in `substituteParams` noting that the order matters for now, but eventually I'd like to change it to just recursively evaluate macros rather than depending on a specific order (it can just error out if it recurses too deep).

What other macros expand to text which should be recursively substituted?

Reply:
I've updated it to substitute user, char, and group after {{description}}. Apologies for the delay.